Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/question/GroupingQuestion.kt

```diff
@@ -0,0 +1,281 @@
+    possibleAnswer: PossibleAnswer.Group,
+    answer: Answer.Group?,
+    onAnswerReady: (Map<String, List<Badge>>) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val initialBadgeList =
+        stringArrayResource(possibleAnswer.items).mapIndexed { index, item ->
+            Badge(index, item)
+        }
+    val badgeGroupList =
+        stringArrayResource(possibleAnswer.groups)
+            // Make sure no group name uses reserved name
+            .filter { it != STARTING_ZONE_BADGE_GROUP }
+            .distinct()
+    var groupAnswer by remember {
+        mutableStateOf(
+            answer?.answerGrouping ?: mapOf()
+        )
+    }
+    val onGroupAnswer: (Pair<String, List<Badge>>) -> Unit = { it ->
+        groupAnswer = groupAnswer.plus(it)
+    }
+    Draggable(modifier.fillMaxSize()) {
+        Column {
+            BadgeGroupStartingZone(initialBadgeList, groupAnswer, onGroupAnswer, onAnswerReady)
+            badgeGroupList.forEach { badgeGroup ->
+                BadgeGroupDropZone(
+                    badgeGroup = badgeGroup,
+                    groupAnswer = groupAnswer,
+                    onGroupAnswer = onGroupAnswer,
+                    modifier = Modifier.padding(vertical = 8.dp),
+                )
+            }
+        }
+    }
+}
+
+data class Badge(val id: Int, val text: String)
+
+@Composable
+private fun BadgeGroupStartingZone(
+    initialBadgeList: List<Badge>,
+    groupAnswer: Map<String, List<Badge>>,
+    onGroupAnswer: (Pair<String, List<Badge>>) -> Unit,
+    onAnswerReady: (Map<String, List<Badge>>) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val isStartingZoneEmpty by remember(groupAnswer) {
+        derivedStateOf { groupAnswer[STARTING_ZONE_BADGE_GROUP]?.isEmpty() == true }
+    }
+    if (isStartingZoneEmpty) {
+        onAnswerReady(groupAnswer)
+    }
+    var badgeList by remember {
+        mutableStateOf(
+            groupAnswer[STARTING_ZONE_BADGE_GROUP] ?: initialBadgeList
+        )
+    }
+    val onItemDropped: (Badge) -> Unit = {
+        badgeList = badgeList.minus(it)
+    }
+    LaunchedEffect(badgeList) {
+        onGroupAnswer(Pair(STARTING_ZONE_BADGE_GROUP, badgeList))
+    }
+
+    DropZone(modifier = modifier, isDropZone = false) { _, _ ->
+        if (badgeList.isNotEmpty()) {
+            BadgeContainer(
+                items = badgeList,
+                contentPadding = PaddingValues(16.dp),
+                onItemDropped = onItemDropped
+            )
+        }
+    }
+}
+
+@Composable
+private fun BadgeGroupDropZone(
+    badgeGroup: String,
+    groupAnswer: Map<String, List<Badge>>,
+    onGroupAnswer: (Pair<String, List<Badge>>) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    var badgeList by remember {
+        mutableStateOf(
+            groupAnswer[badgeGroup] ?: emptyList()
+        )
+    }
+    val onItemDropped: (Badge) -> Unit = {
+        badgeList = badgeList.minus(it)
+    }
+
+    val state = LocalDragTargetInfo.current
```

Comment:
I think you could wrap a lot of this code up in a state holder, it would remove the need for the callbacks and the composition local I believe. It looks like you are using Composition as a way to call the callback which isn't quite right, you should have your state hoisted high enough that when it's changed the relevant components recompose.